File: yt_uploader/uploader.py
"""YouTube auto-uploader watcher.

Watches queue/yt/upload for finished videos (mp4/webm/mov/...). For each
video it finds the matching sidecar (<base>.txt or <base>.description,
written by the machine as "title\n\n<description>"), builds the final
title/description with the required hashtags, verifies the file is complete
(size stability + ffmpeg decode), uploads to YouTube, and moves the pair
into queue/yt/done. Permanent failures move the pair into queue/yt/failed
with an error note. A processed.json record prevents accidental duplicates.

Usage:
  python yt_uploader/uploader.py              watch continuously
  python yt_uploader/uploader.py --once       scan once and exit (testing)
  python yt_uploader/uploader.py --setup-oauth
"""

import argparse
import json
import shutil
import subprocess
import sys
import time
from pathlib import Path

if __package__ in (None, ""):
    sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
    from yt_uploader.config import (
        DONE_DIR, FAILED_DIR, MISSING_SIDECAR_LIMIT, POLL_INTERVAL, SIDECAR_EXTENSIONS,
        STABILITY_DELAY, STATE_PATH, UPLOAD_DELAY, VIDEO_EXTENSIONS, WATCH_DIR,
    )
    from yt_uploader.hashtags import build_description, build_title
    from yt_uploader.logger import configure, fail, log, warn
    from yt_uploader.oauth import OAuthError, get_credentials, run_oauth_flow, token_summary
    from yt_uploader.upload import upload_video
else:
    from .config import (
        DONE_DIR,
        FAILED_DIR,
        MISSING_SIDECAR_LIMIT,
        POLL_INTERVAL,
        SIDECAR_EXTENSIONS,
        STABILITY_DELAY,
        STATE_PATH,
        UPLOAD_DELAY,
        VIDEO_EXTENSIONS,
        WATCH_DIR,
    )
    from .hashtags import build_description, build_title
    from .logger import configure, fail, log, warn
    from .oauth import OAuthError, get_credentials, run_oauth_flow, token_summary
    from .upload import upload_video


def load_state() -> dict:
    if STATE_PATH.exists():
        try:
            return json.loads(STATE_PATH.read_text(encoding="utf-8"))
        except Exception as exc:
            warn(f"State file unreadable ({exc}); starting fresh")
    return {"processed": {}}


def save_state(state: dict):
    STATE_PATH.parent.mkdir(parents=True, exist_ok=True)
    tmp = STATE_PATH.with_suffix(".json.tmp")
    tmp.write_text(json.dumps(state, indent=2, ensure_ascii=False), encoding="utf-8")
    tmp.replace(STATE_PATH)


def scan_videos() -> list:
    if not WATCH_DIR.exists():
        warn(f"Watch dir missing: {WATCH_DIR}")
        return []
    videos = []
    for entry in WATCH_DIR.iterdir():
        if entry.is_file() and not entry.name.startswith("."):
            if entry.suffix.lower() in VIDEO_EXTENSIONS:
                videos.append(entry)
    return sorted(videos)


def find_sidecar(video: Path) -> Path | None:
    for ext in SIDECAR_EXTENSIONS:
        candidate = video.with_suffix(ext)
        if candidate.exists() and candidate.is_file():
            return candidate
    return None


def read_sidecar(sidecar: Path) -> tuple:
    """Return (title, description) from "title\n\n<description>"."""
    text = sidecar.read_text(encoding="utf-8", errors="replace")
    lines = text.split("\n")
    title = lines[0].strip() if lines else ""
    description = "\n".join(lines[1:]).strip()
    return title, description


def is_stable(path: Path) -> bool:
    """True when the file size holds constant across two samples and is > 0."""
    try:
        size_a = path.stat().st_size
        if size_a <= 0:
            return False
        time.sleep(STABILITY_DELAY)
        size_b = path.stat().st_size
        return size_b == size_a
    except OSError:
        return False


def verify_video(path: Path) -> tuple:
    """Full decode check (same approach as machine.js verifyFinalFile).
    Returns (ok, detail)."""
    try:
        proc = subprocess.run(
            ["ffmpeg", "-v", "error", "-i", str(path), "-f", "null", "-"],
            capture_output=True, text=True, timeout=300,
        )
        if proc.returncode != 0:
            return False, (proc.stderr or "ffmpeg decode failed").strip().splitlines()[-1][-300:]
        proc = subprocess.run(
            ["ffprobe", "-v", "error", "-show_entries", "format=duration,size",
             "-show_entries", "stream=codec_type", "-of", "json", str(path)],
            capture_output=True, text=True, timeout=120,
        )
        info = json.loads(proc.stdout or "{}")
        has_video = any(s.get("codec_type") == "video" for s in info.get("streams", []))
        has_audio = any(s.get("codec_type") == "audio" for s in info.get("streams", []))
        duration = float(info.get("format", {}).get("duration", 0) or 0)
        if not has_video or not has_audio or duration <= 0:
            return False, f"bad streams (video={has_video} audio={has_audio} duration={duration})"
        return True, f"{duration:.1f}s"
    except FileNotFoundError:
        warn("ffmpeg/ffprobe not found; skipping decode verification")
        return True, "unverified (no ffmpeg)"
    except subprocess.TimeoutExpired:
        return False, "decode timed out"
    except Exception as exc:
        return False, str(exc)


def move_pair(video: Path, sidecar: Path | None, dest_dir: Path, note: str = None):
    dest_dir.mkdir(parents=True, exist_ok=True)
    if note:
        (dest_dir / f"{video.stem}.error.txt").write_text(note, encoding="utf-8")
    shutil.move(str(video), str(dest_dir / video.name))
    if sidecar and sidecar.exists():
        shutil.move(str(sidecar), str(dest_dir / sidecar.name))


def process_video(video: Path, sidecar: Path | None, state: dict) -> bool:
    base = video.stem
    log(f"Processing {video.name}")

    title, description = ("", "")
    if sidecar:
        try:
            title, description = read_sidecar(sidecar)
        except Exception as exc:
            fail(f"Could not read sidecar {sidecar.name}: {exc}")
            move_pair(video, sidecar, FAILED_DIR, f"unreadable sidecar: {exc}")
            return False
    else:
        warn(f"No sidecar for {video.name}; using derived title")

    final_title = build_title(title or base)
    final_description = build_description(description, base)
    log(f"  Title: {final_title}")
    log(f"  Description: {final_description}")

    ok, detail = verify_video(video)
    if not ok:
        fail(f"{video.name} failed verification: {detail}")
        move_pair(video, sidecar, FAILED_DIR, f"verification failed: {detail}")
        return False

    try:
        url = upload_video(video, final_title, final_description)
    except Exception as exc:
        fail(f"Upload failed for {video.name}: {exc}")
        move_pair(video, sidecar, FAILED_DIR, f"upload failed: {exc}")
        return False

    state.setdefault("processed", {})[base] = {
        "url": url,
        "video": video.name,
        "title": final_title,
        "uploaded_at": time.strftime("%Y-%m-%dT%H:%M:%S"),
    }
    save_state(state)
    move_pair(video, sidecar, DONE_DIR)
    log(f"DONE: {video.name} -> {url} (moved to queue/yt/done)")
    return True


def run_cycle(state: dict, missing_since: dict, delay: float = UPLOAD_DELAY, quiet: bool = False) -> dict:
    """Scan and upload everything pending in the watch dir.

    Returns {"missing_since": dict, "processed": int, "failed": int}.
    When ``quiet`` is true (previous cycle found nothing) an empty scan is
    not logged at all, so a long-running watcher does not spam the console.
    """
    videos = scan_videos()
    if not videos:
        if not quiet:
            log("Scanning queue/yt/upload... no videos pending")
        return {"missing_since": missing_since, "processed": 0, "failed": 0}

    if quiet:
        log("New videos detected in queue/yt/upload")
    log(f"Scanning queue/yt/upload... {len(videos)} video(s) pending")

    processed = 0
    failed = 0
    for index, video in enumerate(videos):
        base = video.stem
        if base in state.get("processed", {}):
            warn(f"{video.name} already processed (recorded); moving to done")
            move_pair(video, find_sidecar(video), DONE_DIR)
            continue

        sidecar = find_sidecar(video)
        if sidecar is None:
            seen = missing_since.get(base, 0) + 1
            missing_since[base] = seen
            if seen < MISSING_SIDECAR_LIMIT:
                log(f"  {video.name}: waiting for sidecar (cycle {seen}/{MISSING_SIDECAR_LIMIT})")
                continue
            warn(f"{video.name}: sidecar still missing after {seen} cycles; proceeding without it")
        else:
            missing_since.pop(base, None)

        if not is_stable(video):
            warn(f"{video.name} still being written; deferring")
            continue

        processed += 1
        if not process_video(video, sidecar, state):
            failed += 1



        if delay > 0 and index < len(videos) - 1:
            log(f"Waiting {delay:.0f} seconds before next upload")
            time.sleep(delay)

    for base in list(missing_since):
        if base not in {v.stem for v in videos}:
            missing_since.pop(base, None)
    return {"missing_since": missing_since, "processed": processed, "failed": failed}


def watch_forever(delay: float = UPLOAD_DELAY):
    state = load_state()
    missing_since = {}
    quiet_cycle = False
    log("=" * 60)
    log("YouTube auto-uploader started")
    log(f"  Watch dir:    {WATCH_DIR}")
    log(f"  Done dir:     {DONE_DIR}")
    log(f"  Failed dir:   {FAILED_DIR}")
    log(f"  Poll interval: {POLL_INTERVAL}s")
    if delay > 0:
        log(f"  Inter-video delay: {delay:.0f}s")
    log(f"  OAuth:        {token_summary()}")
    log("=" * 60)
    get_credentials()

    while True:
        try:
            result = run_cycle(state, missing_since, delay, quiet=quiet_cycle)
            missing_since = result["missing_since"]
            quiet_cycle = result["processed"] == 0 and result["failed"] == 0
        except OAuthError as exc:
            fail(f"OAuth problem: {exc}")
        except Exception as exc:
            fail(f"Cycle error: {exc}")
        time.sleep(POLL_INTERVAL)


def main(argv=None):
    parser = argparse.ArgumentParser(description="YouTube auto-uploader for vs auto")
    parser.add_argument("--once", action="store_true", help="scan once and exit (testing)")
    parser.add_argument("--setup-oauth", action="store_true", help="run one-time OAuth and exit")
    parser.add_argument("--delay", type=float, default=None, help="cooldown between uploads (seconds)")
    parser.add_argument("--verbose", action="store_true", help="debug logging")
    args = parser.parse_args(argv)

    if sys.platform == "win32":
        sys.stdout.reconfigure(encoding="utf-8", errors="replace")
        sys.stderr.reconfigure(encoding="utf-8", errors="replace")
    configure(verbose=args.verbose)

    delay = args.delay if args.delay is not None else UPLOAD_DELAY
    if args.delay is not None:
        log(f"Inter-video delay: {delay:.0f}s (from --delay)")

    if args.setup_oauth:
        run_oauth_flow()
        log(f"Authorization complete. Token: {token_summary()}")
        return 0

    if args.once:
        state = load_state()
        result = run_cycle(state, {}, delay)
        save_state(state)
        if result["failed"]:
            fail(f"Cycle finished: {result['processed']} processed, {result['failed']} failed")
            return 1
        return 0

    watch_forever(delay)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tiktok_uploader/uploader.js

const fs = require("fs/promises");
const path = require("path");
const crypto = require("crypto");
const { execFile } = require("child_process");
const { config } = require("./config");
const { SessionError, UploadError } = require("./errors");
const { assertLoggedIn } = require("./session");



function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function baseName(fileName) {
  const ext = path.extname(fileName);
  return fileName.slice(0, fileName.length - ext.length);
}

function isVideoFile(fileName) {
  return config.videoExtensions.includes(path.extname(fileName).toLowerCase());
}



async function scanVideos(logger) {
  try {
    await fs.access(config.uploadDir);
  } catch {
    logger.warn(`Upload dir missing: ${config.uploadDir}`);
    return [];
  }
  const entries = await fs.readdir(config.uploadDir, { withFileTypes: true });
  const videos = entries
    .filter((e) => e.isFile() && !e.name.startsWith(".") && isVideoFile(e.name))
    .map((e) => e.name)
    .sort();
  return videos;
}

async function findSidecar(videoName) {
  const base = baseName(videoName);
  for (const ext of config.sidecarExtensions) {
    const candidate = path.join(config.uploadDir, `${base}${ext}`);
    try {
      const stat = await fs.stat(candidate);
      if (stat.isFile()) return candidate;
    } catch {

    }
  }
  return null;
}

async function readSidecar(sidecarPath) {
  return (await fs.readFile(sidecarPath, "utf8")).trim();
}

async function isStable(videoName) {
  const full = path.join(config.uploadDir, videoName);
  try {
    const sizeA = (await fs.stat(full)).size;
    if (sizeA <= 0) return false;
    await sleep(config.stabilityDelaySeconds * 1000);
    const sizeB = (await fs.stat(full)).size;
    return sizeB === sizeA && sizeB > 0;
  } catch {
    return false;
  }
}

function runFfmpegDecodeCheck(filePath) {
  return new Promise((resolve) => {
    execFile(
      "ffmpeg",
      ["-v", "error", "-i", filePath, "-f", "null", "-"],
      { maxBuffer: 64 * 1024 * 1024, timeout: 300000 },
      (error, stdout, stderr) => {
        if (error) {
          resolve({ ok: false, detail: String(stderr || error.message).trim().split("\n").pop().slice(-300) });
          return;
        }
        resolve({ ok: true, detail: "decode ok" });
      }
    );
  });
}


async function verifyVideo(videoName, logger) {
  const full = path.join(config.uploadDir, videoName);
  const decode = await runFfmpegDecodeCheck(full);
  if (!decode.ok) {
    return { ok: false, detail: decode.detail };
  }
  return new Promise((resolve) => {
    execFile(
      "ffprobe",
      ["-v", "error", "-show_entries", "format=duration", "-show_entries", "stream=codec_type", "-of", "json", full],
      { maxBuffer: 16 * 1024 * 1024, timeout: 120000 },
      (error, stdout, stderr) => {
        if (error) {
          if (error.code === "ENOENT") {
            logger.warn("ffprobe not found; skipping stream verification");
            resolve({ ok: true, detail: "unverified (no ffprobe)" });
            return;
          }
          resolve({ ok: false, detail: String(stderr || error.message).slice(-300) });
          return;
        }
        try {
          const info = JSON.parse(stdout || "{}");
          const streams = info.streams || [];
          const hasVideo = streams.some((s) => s.codec_type === "video");
          const hasAudio = streams.some((s) => s.codec_type === "audio");
          const duration = Number(info.format?.duration || 0);
          if (!hasVideo || !hasAudio || duration <= 0) {
            resolve({ ok: false, detail: `bad streams (video=${hasVideo} audio=${hasAudio} duration=${duration})` });
            return;
          }
          resolve({ ok: true, detail: `${duration.toFixed(1)}s` });
        } catch (exc) {
          resolve({ ok: false, detail: `ffprobe output unreadable: ${exc.message}` });
        }
      }
    );
  });
}



async function loadState(logger) {
  try {
    return JSON.parse(await fs.readFile(config.statePath, "utf8"));
  } catch {
    return { processed: {} };
  }
}

async function saveState(state) {
  await fs.mkdir(config.stateDir, { recursive: true });
  const tmp = `${config.statePath}.tmp-${crypto.randomBytes(4).toString("hex")}`;
  await fs.writeFile(tmp, JSON.stringify(state, null, 2), "utf8");
  await fs.rename(tmp, config.statePath);
}



function pidAlive(pid) {
  if (!pid || !Number.isInteger(pid) || pid <= 0) return false;
  try {
    process.kill(pid, 0);
    return true;
  } catch (error) {
    return error.code === "EPERM";
  }
}


async function acquireLock(logger) {
  await fs.mkdir(config.stateDir, { recursive: true });
  const entry = { pid: process.pid, startedAt: new Date().toISOString() };
  try {
    const handle = await fs.open(config.lockPath, "wx");
    await handle.writeFile(JSON.stringify(entry));
    await handle.close();
    logger.log(`Lock acquired (pid ${process.pid})`);
    return true;
  } catch (error) {
    if (error.code !== "EEXIST") throw error;
    let stale = true;
    try {
      const existing = JSON.parse(await fs.readFile(config.lockPath, "utf8"));
      stale = !pidAlive(existing.pid);
    } catch {

    }
    if (stale) {
      logger.warn("Stale lock detected (previous process died); reclaiming");
      await fs.rm(config.lockPath, { force: true });
      return acquireLock(logger);
    }
    return false;
  }
}

async function releaseLock() {
  await fs.rm(config.lockPath, { force: true });
}



async function moveToDone(videoName, sidecarPath, logger) {
  await fs.mkdir(config.doneDir, { recursive: true });
  const srcVideo = path.join(config.uploadDir, videoName);
  const dstVideo = path.join(config.doneDir, videoName);

  const videoExists = await fs
    .access(srcVideo)
    .then(() => true)
    .catch(() => false);
  if (!videoExists) {

    logger.warn(`${videoName} already gone from upload folder; skipping move`);
    return;
  }
  await fs.rename(srcVideo, dstVideo);
  if (sidecarPath) {
    const sidecarName = path.basename(sidecarPath);
    const exists = await fs
      .access(sidecarPath)
      .then(() => true)
      .catch(() => false);
    if (exists) {
      await fs.rename(sidecarPath, path.join(config.doneDir, sidecarName));
    }
  }
  logger.ok(`Moved ${videoName}${sidecarPath ? ` + ${path.basename(sidecarPath)}` : ""} -> queue/tiktok/done`);
}



async function waitForAny(page, selectors, timeoutMs, description, { attachedOnly = false } = {}) {
  const deadline = Date.now() + timeoutMs;
  let lastError = null;
  while (Date.now() < deadline) {
    for (const selector of selectors) {
      try {
        const locator = page.locator(selector).first();
        if (attachedOnly) {
          if ((await locator.count()) > 0) {
            return locator;
          }
        } else if (await locator.isVisible({ timeout: 500 })) {
          return locator;
        }
      } catch (error) {
        lastError = error;
      }
    }
    await sleep(500);
  }
  throw new UploadError(
    `Timed out waiting for ${description} (${timeoutMs / 1000}s). ` +
      `Last error: ${lastError ? lastError.message.split("\n")[0] : "none"}`
  );
}

async function findPostButton(page) {
  const deadline = Date.now() + 30000;
  while (Date.now() < deadline) {
    for (const selector of config.selectors.postButton) {
      try {
        const locator = page.locator(selector).first();
        if (await locator.isVisible({ timeout: 500 })) {
          const disabled = await locator.isDisabled().catch(() => false);
          if (!disabled) return locator;
        }
      } catch {

      }
    }
    await sleep(500);
  }
  return null;
}

async function removeOverlays(page) {



  try {
    await page.evaluate(() => {
      const portal = document.getElementById("react-joyride-portal");
      if (portal) portal.remove();
    });
    await sleep(300);
  } catch {

  }
}

async function guardedClick(page, locator, description) {
  for (let attempt = 0; attempt < 3; attempt += 1) {
    await removeOverlays(page);
    try {
      await locator.click({ timeout: 8000 });
      return;
    } catch (error) {
      if (attempt === 2) {
        throw new UploadError(`Could not click ${description}: ${error.message.split("\n")[0]}`);
      }
    }
  }
}

async function fillCaption(page, caption) {
  const box = await waitForAny(page, config.selectors.captionBox, 120000, "caption box");
  await removeOverlays(page);
  const text = String(caption || "");


  await guardedClick(page, box, "caption box");
  await page.keyboard.press("ControlOrMeta+a");
  await page.keyboard.type(text, { delay: 2 });
  await sleep(500);
  const typed = await box.textContent().catch(() => "");
  if (text.trim() && !(typed || "").includes(text.slice(0, 20))) {
    await guardedClick(page, box, "caption box");
    await page.keyboard.press("ControlOrMeta+a");
    await page.keyboard.type(text, { delay: 5 });
    await sleep(500);
  }
  return box;
}

async function waitForUploadProgress(page, logger) {

  await waitForAny(page, config.selectors.progressBar, 60000, "upload progress indicator").catch(() => null);

  const deadline = Date.now() + config.uploadTimeoutMs;
  let lastReported = 0;
  while (Date.now() < deadline) {
    const postButton = await findPostButton(page);
    if (postButton) {
      logger.log("Upload progress: complete (file processed, Post available)");
      return postButton;
    }
    try {
      const bar = page.locator(config.selectors.progressBar[0]).first();
      if (await bar.isVisible({ timeout: 300 })) {
        const pct = await bar.getAttribute("aria-valuenow").catch(() => null);
        if (pct) {
          const value = Math.round(Number(pct));
          if (value > lastReported && value % 25 === 0) {
            logger.log(`Upload progress: ${value}%`);
            lastReported = value;
          }
        }
      }
    } catch {

    }
    await sleep(2000);
  }
  throw new UploadError(`Timed out waiting for the upload to finish (${config.uploadTimeoutMs / 1000}s)`);
}

async function waitForSuccess(page, logger) {
  const deadline = Date.now() + config.confirmTimeoutMs;
  while (Date.now() < deadline) {

    for (const selector of config.selectors.successSignals) {
      try {
        const locator = page.locator(selector).first();
        if (await locator.isVisible({ timeout: 400 })) {
          logger.ok("Success dialog detected");
          return true;
        }
      } catch {

      }
    }


    if (!/tiktokstudio\/upload/i.test(page.url())) {
      logger.ok("TikTok redirected to posts page — upload confirmed");
      return true;
    }

    let errorText = null;
    try {
      const errorToast = page
        .locator("text=Upload failed")
        .or(page.locator("text=Posting failed"))
        .or(page.locator("text=Something went wrong"))
        .first();
      if (await errorToast.isVisible({ timeout: 400 })) {
        errorText = "TikTok reported an upload error toast.";
      }
    } catch {

    }
    if (errorText) {
      throw new UploadError(errorText);
    }
    await sleep(2000);
  }
  throw new UploadError(`Timed out waiting for TikTok to confirm the post (${config.confirmTimeoutMs / 1000}s)`);
}

async function extractPostedUrl(page) {
  try {
    const link = page
      .locator('a[href*="/video/"]')
      .or(page.locator("a:has-text('View video')"))
      .first();
    if (await link.isVisible({ timeout: 1500 })) {
      return link.getAttribute("href").catch(() => null);
    }
  } catch {

  }
  return null;
}


async function uploadVideo({ context, videoPath, caption, logger }) {
  const page = await context.newPage();
  try {
    logger.log("Upload started");
    await page.goto(config.uploadUrl, {
      waitUntil: "domcontentloaded",
      timeout: config.navTimeoutMs,
    });
    await page.waitForLoadState("networkidle", { timeout: 30000 }).catch(() => {});
    await assertLoggedIn(page, 20000);

    const fileInput = await waitForAny(page, config.selectors.fileInput, 30000, "file input", {
      attachedOnly: true,
    });
    await fileInput.setInputFiles(videoPath);
    logger.ok(`File attached: ${path.basename(videoPath)}`);

    await fillCaption(page, caption);
    logger.ok("Caption loaded");

    const postButton = await waitForUploadProgress(page, logger);
    await guardedClick(page, postButton, "Post button");
    logger.log("Post clicked — waiting for TikTok confirmation");

    await waitForSuccess(page, logger);
    const url = await extractPostedUrl(page);
    logger.ok("Upload completed — TikTok confirmed the post");
    return { url };
  } catch (error) {
    const screenshot = path.join(config.stateDir, `debug-${Date.now()}.png`);
    await page.screenshot({ path: screenshot, fullPage: true }).catch(() => {});
    if (error instanceof SessionError || error instanceof UploadError) {
      logger.warn(`Debug screenshot saved: ${screenshot}`);
      throw error;
    }
    logger.warn(`Debug screenshot saved: ${screenshot}`);
    throw new UploadError(`TikTok upload page error: ${error.message.split("\n")[0]}`, { cause: error });
  } finally {
    await page.close().catch(() => {});
  }
}




async function processVideo(videoName, sidecarPath, state, { context, logger }) {
  const base = baseName(videoName);
  const record = {
    base,
    video: videoName,
    sidecar: sidecarPath ? path.basename(sidecarPath) : null,
    status: "failed",
    attempts: 0,
    error: null,
  };


  if (state.processed[base]) {
    logger.warn(`${videoName} already processed (recorded); moving to done`);
    await moveToDone(videoName, sidecarPath, logger);
    return { ...record, status: "already-posted" };
  }


  let caption = "";
  if (sidecarPath) {
    try {
      caption = await readSidecar(sidecarPath);
      logger.log(`Description loaded (${caption.length} chars)`);
    } catch (error) {
      logger.fail(`Could not read ${path.basename(sidecarPath)}: ${error.message}`);
    }
  }

  const verified = await verifyVideo(videoName, logger);
  if (!verified.ok) {
    record.error = `video failed verification: ${verified.detail}`;
    logger.fail(`${videoName} — ${record.error} (kept in upload folder)`);
    return record;
  }
  logger.log(`${videoName} verified (${verified.detail})`);

  const attempts = config.maxRetries + 1;
  let lastError = null;
  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    record.attempts = attempt;
    try {
      const result = await uploadVideo({
        context,
        videoPath: path.join(config.uploadDir, videoName),
        caption,
        logger,
      });

      state.processed[base] = {
        url: result.url,
        video: videoName,
        uploadedAt: new Date().toISOString(),
      };
      await saveState(state);
      await moveToDone(videoName, sidecarPath, logger);
      record.status = "ok";
      record.url = result.url;
      logger.ok(`DONE: ${videoName} posted${result.url ? ` -> ${result.url}` : ""}`);
      return record;
    } catch (error) {
      lastError = error;
      if (error instanceof SessionError) {
        record.error = error.message;
        logger.fail(`${videoName} — ${error.message}`);
        return record;
      }
      if (attempt < attempts) {
        const backoff = config.retryBackoffMs[Math.min(attempt - 1, config.retryBackoffMs.length - 1)];
        logger.warn(
          `${videoName} upload attempt ${attempt}/${attempts} failed (${error.message}); retrying in ${backoff / 1000}s`
        );
        await sleep(backoff);
      }
    }
  }

  record.error = lastError ? lastError.message : "unknown error";
  logger.fail(`${videoName} failed after ${attempts} attempts: ${record.error} (kept in upload folder)`);
  return record;
}


async function runCycle({ context, state, logger, missingSidecarSince = {}, processVideo: processVideoImpl = processVideo, quiet = false }) {
  state.processed = state.processed || {};
  const videos = await scanVideos(logger);
  if (videos.length === 0) {
    if (!quiet) {
      logger.log("Scanning queue/tiktok/upload... no videos pending");
    }
    return { processed: [], missingSidecarSince, videos: 0 };
  }

  if (quiet) {
    logger.log("New videos detected in queue/tiktok/upload");
  }
  logger.log(`Scanning queue/tiktok/upload... ${videos.length} video(s) pending`);
  const processed = [];

  for (let index = 0; index < videos.length; index += 1) {
    const videoName = videos[index];
    const base = baseName(videoName);
    let sidecarPath = await findSidecar(videoName);

    if (!sidecarPath) {
      const seen = (missingSidecarSince[base] || 0) + 1;
      missingSidecarSince[base] = seen;
      if (seen < config.missingSidecarWaitCycles) {
        logger.log(`${videoName}: no sidecar yet (cycle ${seen}/${config.missingSidecarWaitCycles}); deferring`);
        continue;
      }
      logger.warn(`${videoName}: no sidecar after ${seen} cycles; proceeding with empty caption`);
    } else {
      delete missingSidecarSince[base];
    }

    if (!(await isStable(videoName))) {
      logger.warn(`${videoName} still being written; deferring`);
      continue;
    }

    logger.log(`Uploading: ${videoName}`);
    processed.push(await processVideoImpl(videoName, sidecarPath, state, { context, logger }));



    if (config.delaySeconds > 0 && index < videos.length - 1) {
      logger.log(`Waiting ${config.delaySeconds} seconds before next upload`);
      await sleep(config.delaySeconds * 1000);
    }
  }

  for (const base of Object.keys(missingSidecarSince)) {
    if (!videos.includes(`${base}.mp4`) && !videos.some((v) => baseName(v) === base)) {
      delete missingSidecarSince[base];
    }
  }

  return { processed, missingSidecarSince, videos: videos.length };
}

module.exports = {
  scanVideos,
  findSidecar,
  readSidecar,
  isStable,
  verifyVideo,
  loadState,
  saveState,
  acquireLock,
  releaseLock,
  moveToDone,
  uploadVideo,
  waitForSuccess,
  processVideo,
  runCycle,
  baseName,
};
